lib/pos_transaction_register.go

```diff
@@ -186,6 +249,91 @@ func (tr *TransactionRegister) GetFeeTimeTransactions() []*MempoolTx {
 	return txns
 }
 
+// PruneToSize removes transactions from the end of the register until the size of the register shrinks to the desired
+// number. The returned transactions are ordered by lowest-to-highest priority, i.e. first transaction will have the
+// smallest fee, last transaction will have the highest fee. Returns no error if no transactions were pruned.
+func (tr *TransactionRegister) PruneToSize(maxSizeBytes uint64) (_prunedTxns []*MempoolTx, _err error) {
+	tr.Lock()
+	defer tr.Unlock()
+
+	// If the maximum number of bytes is greater or equal to the current size of the register, return.
+	if maxSizeBytes >= tr.totalTxnsSizeBytes {
+		return nil, nil
+	}
+
+	// If the register is empty, return.
+	if tr.feeTimeBucketSet.Empty() {
+		return nil, nil
+	}
+
+	// Determine how many bytes we need to prune and get the transactions to prune.
+	minPrunedBytes := tr.totalTxnsSizeBytes - maxSizeBytes
+	prunedTxns, err := tr.getTransactionsToPrune(minPrunedBytes)
+	if err != nil {
+		return nil, errors.Wrapf(err, "TransactionRegister.PruneToSize: Error getting transactions to prune")
+	}
+
+	// Remove the transactions from prunedTxns.
+	for _, txn := range prunedTxns {
+		if err := tr.removeTransactionNoLock(txn); err != nil {
+			return nil, errors.Wrapf(err, "TransactionRegister.PruneToSize: Error removing transaction %v", txn.Hash.String())
+		}
+	}
+	return prunedTxns, nil
+}
+
+func (tr *TransactionRegister) getTransactionsToPrune(minPrunedBytes uint64) (_prunedTxns []*MempoolTx, _err error) {
+	if minPrunedBytes == 0 {
+		return nil, nil
+	}
+
+	prunedBytes := uint64(0)
+	prunedTxns := []*MempoolTx{}
+
+	// Find the FeeTime bucket at the end of the Set. It'll have the smallest fee among the buckets in the register.
+	it := tr.feeTimeBucketSet.Iterator()
+	it.End()
+	// Iterate through the buckets in reverse order so that we drop transactions ordered by least-to-highest priority.
+	for it.Prev() {
+		bucket, ok := it.Value().(*FeeTimeBucket)
+		if !ok {
+			return nil, fmt.Errorf("TransactionRegister.getTransactionsToPrune: " +
+				"Error casting value of FeeTimeBucket")
+		}
+
+		// Iterate through the transactions in the current FeeTime bucket. We iterate in reverse order, starting from the
+		// end, so that we drop transactions ordered by least-to-highest priority.
+		bucketIt := bucket.GetIterator()
+		bucketIt.End()
+		for bucketIt.Prev() {
+			txn, ok := bucketIt.Value().(*MempoolTx)
+			if !ok {
+				return nil, fmt.Errorf("TransactionRegister.getTransactionsToPrune: " +
+					"Error casting value of MempoolTx")
+			}
+			// Make sure that, somehow, the number of pruned bytes doesn't overflow uint64.
+			if prunedBytes > math.MaxUint64-txn.TxSizeBytes {
+				return nil, fmt.Errorf("TransactionRegister.getTransactionsToPrune: "+
+					"Error pruning %v bytes from register: prunedBytes %v exceeds max uint64", minPrunedBytes, prunedBytes)
+			}
```

Comment:
IMO you can remove this consistency check. Preventing size uint64 overflow is one of guarantees of the TransactionRegister struct. And you validate that each txn doesn't cause a uint64 overflow of the TransactionRegister's size before adding it to the register. Since all those checks are encapsulated in this struct, it's safe to assume here that this condition can never be hit.
```suggestion
```